com.unity.multiplayer.mlapi/Runtime/Core/MessageBatcher.cs

```diff
@@ -0,0 +1,198 @@
+using MLAPI.Serialization.Pooled;
+using MLAPI.Serialization;
+using MLAPI.Configuration;
+using MLAPI.Profiling;
+using System.Collections.Generic;
+using System.Linq;
+using System;
+using System.IO;
+
+namespace MLAPI
+{
+    class MessageBatcher
+    {
+        public class SendStream
+        {
+            public FrameQueueItem Item;
+            public PooledBitStream Stream = new PooledBitStream();
+        }
+
+        // Stores the stream of batched RPC to send to each client, by ClientId
+        private Dictionary<ulong, SendStream> SendDict = new Dictionary<ulong, SendStream>();
+        private PooledBitWriter Writer = new PooledBitWriter(new PooledBitStream());
+
+        // Used to store targets, internally
+        private List<ulong> TargetList = new List<ulong>();
+
+        public void PushLength(int length, ref PooledBitWriter writer)
+        {
+            // If length is single byte we write it
+            if (length < 256)
+            {
+                writer.WriteByte((byte)length); // write the amounts of bytes that are coming up
+            }
+            else
+            {
+                // otherwise we write a 0, and a two-byte length
+                writer.WriteByte(0); // mark larger size
+                writer.WriteByte((byte)(length % 256)); // write the length modulo 256
+                writer.WriteByte((byte)(length / 256)); // write the length divided by 256
+            }
+        }
+
+        public int PopLength(in BitStream messageStream)
+        {
+            byte read = (byte)messageStream.ReadByte();
+            // if we read a non-zero value, we have a single byte length
+            if (read != 0)
+            {
+                return read;
+            }
+            // otherwise, a two-byte length follows
+            int length = messageStream.ReadByte();
+            length += messageStream.ReadByte() * 256;
+
+            return length;
+        }
+
+        /// <summary>
+        /// FillTargetList
+        /// Fills a list with the ClientId's an item is targeted to
+        /// </summary>
+        /// <param name="item">the FrameQueueItem we want targets for</param>
+        /// <param name="list">the list to fill</param>
+        private static void FillTargetList(in FrameQueueItem item, ref List<ulong> list)
+        {
+            list.Clear();
+            switch (item.QueueItemType)
+            {
+                case RPCQueueManager.QueueItemType.ServerRpc:
+                    list.Add(item.NetworkId);
+                    break;
+                case RPCQueueManager.QueueItemType.ClientRpc:
+                    foreach (ulong clientId in item.ClientIds)
+                    {
+                        list.Add(clientId);
```

Comment:
kind of dangerous probably, but since list is passed in by ref, could you avoid the copy and just do

list = item.ClientIds 

(assuming the caller won't mutate it)

Reply:
I can do this:

```
-        private List<ulong> TargetList = new List<ulong>();
+        private ulong[] TargetList = new ulong[0];
```

and

```
-        private static void FillTargetList(in FrameQueueItem item, ref List<ulong> list)
+        private static void FillTargetList(in FrameQueueItem item, ref ulong[] list)
         {
-            list.Clear();
             switch (item.QueueItemType)
             {
                 case RPCQueueManager.QueueItemType.ServerRpc:
-                    list.Add(item.NetworkId);
+                    Array.Resize(ref list, 1);
+                    list[0] = item.NetworkId;
                     break;
                 case RPCQueueManager.QueueItemType.ClientRpc:
-                    foreach (ulong clientId in item.ClientIds)
-                    {
-                        list.Add(clientId);
-                    }
+                    list = item.ClientIds;
                     break;
```

It improves the ClientRPC case, I'm not sure about the ServerRPC case. And I can't speak to performance. Which do you prefer?